src/main.c

```diff
@@ -744,6 +766,168 @@ static void on_webctx_init_web_extension(WebKitWebContext *webctx, gpointer data
 }
 
 /**
+ * Callback for the webkit download decide destination signal.
+ * This signal is emitted after response is received to decide a destination
+ * URI for the download.
+ */
+static gboolean on_webdownload_decide_destination(WebKitDownload *download,
+        gchar *suggested_filename, Client *c)
+{
+    g_assert(download);
+    g_assert(suggested_filename);
+    g_assert(c);
+
+    char *path, *filename, *uri;
+    GString *expanded, *filepath;
+    const char *extension_dot;
+    int suffix;
+    gssize suffix_pos;
+
+    /* get the download path from settings */
+    path = GET_CHAR(c, "download-path");
+    g_assert(path);
+
+    /* expand any ~ or $VAR patterns in download path */
+    expanded = g_string_new(NULL);
+    util_parse_expansion(c, (const char**)&path, expanded, UTIL_EXP_TILDE|UTIL_EXP_DOLLAR, "");
+    g_string_append(expanded, path + 1);
+
+    /* for unnamed downloads set default filename */
+    filename = strlen(suggested_filename) ? suggested_filename : "vimb-download";
+
+    /* construct complete download filepath */
+    filepath = g_string_new(NULL);
+    g_string_printf(filepath, "%s%c%s", expanded->str, G_DIR_SEPARATOR, filename);
+
+    /* if the filepath exists already
+     * insert numerical suffix before file extension */
+    if (g_file_test(filepath->str, G_FILE_TEST_EXISTS)) {
+        suffix = 2;
+
+        /* position on .tar. (special case, extension with two dots),
+         * position on last dot (if any) otherwise */
+        if (!(extension_dot = strstr(filename, ".tar."))) {
+            extension_dot = strrchr(filename, '.');
+        }
+
+        /* the position to insert the suffix at */
+        if (extension_dot) {
+            suffix_pos = extension_dot - filename;
+        } else {
+            suffix_pos = strlen(filename);
+        }
+
+        /* construct a new complete download filepath and add the suffix before
+         * the filename extension, keep incrementing the suffix value as long
+         * as the filepath exists, stop on first unused filename. */
+        do {
+            g_string_printf(filepath, "%s%c%.*s_%i%s",
```

Comment:
In my opinion, the filepath have to be freed during looping by `g_free()`. I think, there must be an `g_free(filepath);` before the `g_string_printf`.